frontend/src/components/authentication/UserSignIn.tsx

```diff
@@ -0,0 +1,269 @@
+import { UseFormSetError } from "react-hook-form";
+
+const messages = defineMessages({
+  submit: {
+    id: "UserSignIn.submit",
+    defaultMessage: "Sign In",
+  },
+  backupPasswordRequired: {
+    id: "OpenIdConnectRedirect.signInForm.backupPasswordRequired",
+    defaultMessage: "A backup password is required",
+  },
+  decryptionFailed: {
+    id: "OpenIdConnectRedirect.signInForm.decryptionFailed",
+    defaultMessage: "Decryption failed, try a different password.",
+  },
+  unknownError: {
+    id: "OpenIdConnectRedirect.signInForm.unknownError",
+    defaultMessage: "An unknown error occurred",
+  },
+});
+
+class DecryptionError extends Error {
+  constructor() {
+    super("Decryption failed");
+  }
+}
+
+class NoBockupPasswordError extends Error {
+  constructor() {
+    super("No backup password provided");
+  }
+}
+
+const UserSignIn = ({
+  idToken,
+  redirectPath,
+  authResponse,
+}: {
+  idToken: string;
+  redirectPath: string;
+  authResponse: AuthenticationResponseDto;
+}) => {
+  const intl = useIntl();
+  const router = useRouter();
+  const updateUserKey = useUpdateUserKey();
+  const updateAuthenticationContext = useContext(UpdateAuthenticationContext);
+
+  const tryUserSignIn = useCallback(
+    async (
+      userProvidedPassword: string,
+      userProvidedBackupPassword?: string,
+    ) => {
+      const {
+        id: userId,
+        name: userName,
+        email: userEmail,
+        authenticationToken,
+        keyPair,
+        type,
+      } = authResponse;
+
+      if (
+        authResponse.type !== UserType.ADMIN &&
+        authResponse.type !== UserType.TEACHER
+      ) {
+        throw new Error(`Unsupported user type: ${authResponse.type}`);
+      }
+
+      const crypto = window.crypto.subtle;
+      let teacherKeyPair: TeacherLongTermKeyPair;
+      let keyPairId: number;
+
+      if (keyPair === null) {
+        if (!userProvidedBackupPassword) {
+          throw new NoBockupPasswordError();
+        }
+
+        // this is the first time the teacher is logging in, create a new key pair
+
+        teacherKeyPair = await TeacherLongTermKeyPair.generate(
+          window.crypto.subtle,
+        );
+
+        const salt1 = PasswordDerivedKey.generateSalt();
+        const pwKey1 = await PasswordDerivedKey.derive(
+          crypto,
+          userProvidedPassword,
+          salt1,
+        );
+
+        const salt2 = PasswordDerivedKey.generateSalt();
+        const pwKey2 = await PasswordDerivedKey.derive(
+          crypto,
+          userProvidedBackupPassword,
+          salt2,
+        );
+
+        const privateKey1 = await teacherKeyPair.exportPrivateKey(pwKey1);
+        const privateKey2 = await teacherKeyPair.exportPrivateKey(pwKey2);
+
+        const publicKey = await teacherKeyPair.exportPublicKey();
+        const saltPublicKey = await teacherKeyPair.exportSaltPublicKey();
+        const fingerprint = await teacherKeyPair.getPublicKeyFingerprint();
```

Comment:
should all this stuff (and some more aronud it) be in its own crypto-specific function instead of being in a component?

Reply:
Could be although this is only calling crypto functions in a way specific to the API/application? But happy to factor it out